python/tskit/metadata.py

```diff
@@ -0,0 +1,165 @@
+# MIT License
+#
+# Copyright (c) 2020 Tskit Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Classes for metadata decoding, encoding and validation
+"""
+import abc
+import json
+from typing import Any
+from typing import Optional
+
+import jsonschema
+
+import tskit.exceptions as exceptions
+
+
+class AbstractMetadataSchema(metaclass=abc.ABCMeta):
+    """
+    Superclass of all MetadataSchemas
+    """
+
+    @property
+    @staticmethod
+    @abc.abstractmethod
+    def name() -> str:
+        """
+        String used to identify this encoding
+        """
+        raise NotImplementedError  # pragma: no cover
+
+    def __init__(self, schema: dict) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    def __str__(self) -> str:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def validate_and_encode_row(self, row: Any) -> bytes:
+        """
+        Validate a row of metadata against this schema and return the encoded
+        representation
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def decode_row(self, row: bytes) -> Any:
+        """
+        Decode an encoded row of metadata, note that no validation is performed
+        """
+        raise NotImplementedError
+
+
+class JSONMetadataSchema(AbstractMetadataSchema):
```

Comment:
There's a bit of the design I see I misunderstood now. I thought we would have a class, MetadataSchema which would have a property ``codec`` which is either NullCodec, or JSONCodec, ... 

It's confusing to me that the MetadataSchema class is bound up with the details of the codec. Shouldn't the MetadataSchema class just be responsible for parsing the *schema* and then creating an instance of the correct codec?

Reply:
That's exactly how it worked initially, when you first looked over the other PR. I combined the two classes recently as I realised that different codecs would need different meta-schemas. For example struct needing `format` and `index`. So once `struct` is implemented there would be a `StructMetadataSchema` class. It didn't seem right to have `MetadataSchema` creating a schema and matching codec. I agree the naming is now confusing though as it's a JSON schema and a JSON encoding! We could have a `MetadataSchema` class that delegates both the encoding _and_ the schema to a `(JSON|Struct)Codec` class.

Reply:
I think it's OK if the MetadataSchema class knows about the codecs it supports and the extra fields that are needed for the different codecs. Otherwise I can see this becoming a tangled web of Java-like inheritance, which I really want to avoid.

Basically, the schema is always the same, it's just that we have a couple of extra fields that are required when the codec (this is probably a better name than ``format`` actually) is ``struct``.

Reply:
I've had a go at refactoring in https://github.com/tskit-dev/tskit/pull/543/commits/711bf1fe614b40f791bb356801123a94290e9307 
Now there is only one `MetadataSchema` that uses `***Codec` classes.